src/open-url.coffee

```diff
@@ -30,90 +30,92 @@ module.exports = (robot) ->
+	* (err, res, body) and renders them to a Promise function (resolve, reject)
 	###
 	createResolver = (resolve, reject) ->
 		(err, res, body) ->
 			if (not err?) and (res.statusCode is 200)
 				resolve(body)
 			else
-				reject(err ? new Error("StatusCode: #{res.statusCode}"))
+				reject(err ? new Error("StatusCode: #{res.statusCode}; Body: #{body}"))
 
 	###*
-	* This creates a function to report errors
-	* @param {Object} Hubot msg object
-	* @return {function} Takes (Error) and reports it
+	* Creates a function that communicates an error.
+	* @param {Object} context - A Hubot msg object
+	* @return {function} - a function that takes (error) and communicates it
 	###
 	createErrorReporter = (context) ->
 		(error) ->
 			robot.logger.error(error)
 			context.send('Something went wrong. Debug output logged')
 
 	###*
-	* Attempt to visit the url referenced by key
-	* @param {string} url to visit
-	* @return {Promise} Handle output from the asynchronous
+	* Attempt to finger the url referenced by key using GET
+	* @param {string} url - address of the web site to finger
+	* @return {Promise} - A promise for the response for the given url
 	###
-	open = (url) ->
+	finger = (url) ->
 		new Promise (resolve, reject) ->
 			robot.http(url).get() createResolver(resolve, reject)
 
 	###*
 	* Store a url in the cache and firebase
-	* @param {string} key under which to store it
-	* @param {string} bookmark to store
-	* @return {Promise} Handle output from the asynchronous section
+	* @param {string} namespace - a namespace under which to store the value
+	* @param {string} key - a name for the value
+	* @param {string} value - url to store
+	* @return {Promise} - A promise for the response from storing the given pair
 	###
-	bookmark = (key, value) ->
+	bookmark = (namespace, key, value) ->
 		new Promise (resolve, reject) ->
-			bookmarks[key] = value
+			bookmarks[namespace] ?= {}
+			bookmarks[namespace][key] = value
 			robot.http("#{firebaseUrl}/data/.json?auth=#{firebaseAuth}")
 				.patch(JSON.stringify({ bookmarks: bookmarks })) createResolver(resolve, reject)
 
 	###*
-	* Extract a value from the context provided
-	* @param {Object} Hubot msg object
-	* @return {string} Value from the stored bookmarks
+	* Extract a value from the context provided.
+	* Attempts to use first match then room name.
+	* @param {Object} context - A Hubot msg object
+	* @return {string} - Value from the stored bookmarks
 	###
-	getValueFromContext = (context) ->
-		if bookmarks[context.match[1]]?
-			return bookmarks[context.match[1]]
-		else if bookmarks[context.envelope.room]?
-			return bookmarks[context.envelope.room]
+	getBookmarkFromContext = ({ match: [_, bookmarkName], envelope: { room: roomName } }) ->
+		if bookmarkName?
+			return bookmarks.named[bookmarkName]
+		else if [roomName]?
+			return bookmarks.rooms[roomName]
 		else
-			throw new Error("Couldn't find key.")
+			throw new Error('No key specified.')
```

Comment:
This is better (it doesn't silently do the wrong thing) but now if you ask provide a key, but for a bookmark that doesn't exist, then we'll try to request from `undefined` and you'll probably get a mysterious error back from `robot.http`, because we pass the result of this directly into `finger` below.

We should either check the result of this function later on, or change this method to throw an exception both if you don't pass a key and also if your key doesn't match anything.